components/claim/components/oauth_to_email.tsx

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {useRef, useState} from 'react';
+import {FormattedMessage} from 'react-intl';
+
+import classNames from 'classnames';
+
+import {AuthChangeResponse} from '@mattermost/types/users';
+
+import {oauthToEmail} from 'actions/admin_actions.jsx';
+
+import Constants from 'utils/constants';
+import {t} from 'utils/i18n';
+import {getPasswordConfig, isValidPassword, localizeMessage, toTitleCase} from 'utils/utils';
+
+import LocalizedInput from 'components/localized_input/localized_input';
+
+type Props = {
+    currentType: string | null;
+    email: string | null;
+    siteName?: string;
+    passwordConfig?: ReturnType<typeof getPasswordConfig>;
+}
+
+const OAuthToEmail = (props: Props) => {
+    const passwordInput = useRef<HTMLInputElement>(null);
+    const passwordConfirmInput = useRef<HTMLInputElement>(null);
+
+    const [error, setError] = useState<string | JSX.Element>('');
+
+    const submit = (e: React.FormEvent) => {
+        e.preventDefault();
+
+        const password = passwordInput.current?.value;
+        if (!password) {
+            setError(localizeMessage('claim.oauth_to_email.enterPwd', 'Please enter a password.'));
+            return;
+        }
+
+        if (props.passwordConfig) {
+            const {valid, error} = isValidPassword(password, props.passwordConfig);
+            if (!valid && error) {
+                setError(error);
+                return;
+            }
+        }
+
+        const confirmPassword = passwordConfirmInput.current?.value;
+        if (!confirmPassword || password !== confirmPassword) {
+            setError(localizeMessage('claim.oauth_to_email.pwdNotMatch', 'Passwords do not match.'));
+            return;
+        }
+
+        setError('');
+
+        oauthToEmail(
+            props.currentType,
+            props.email,
+            password,
+            (data: AuthChangeResponse) => {
+                if (data.follow_link) {
+                    window.location.href = data.follow_link;
+                }
+            },
+            (err: {message: string}) => {
+                setError(err.message);
+            },
+        );
+    };
+
+    const errorElement = error ? <div className='form-group has-error'><label className='control-label'>{error}</label></div> : null;
+    const formClass = classNames('form-group', {'has-error': errorElement});
```

Comment:
this can directly go into the props as well

Reply:
Done

Reply:
thats actually pretty inconsistent in how react handles the re-rendering. I know that passing in literals is causing it to re-render, but afaik when calling a function in props (like with `classNames` or handler-factories) it is memoized instead. I will try and find a ressource that pictures that more clearly. I know that we typically add calls to `classNames` directly into the props without worrying about it too much though